doc/Solr-Configuration.md
# Blacklight configuration

The example Blacklight configuration is a simple application for displaying library holdings. In this section, we will describe the Blacklight configuration settings that determine how the Blacklight interface works with your data.

## Connecting to Solr

The Solr connection parameters are configured globally in `config/solr.yml`. It looks something like this:

```yaml
# config/solr.yml
development:
  url: <%= ENV['SOLR_URL'] || "http://127.0.0.1:8983/solr" %>
test: &test
  url: <%= "http://127.0.0.1:#{ENV['TEST_JETTY_PORT'] || 8888}/solr" %>
```

The configuration is used to configure [RSolr::Client](https://github.com/rsolr/rsolr/blob/master/lib/rsolr/client.rb). Available options include:

* url
* proxy
* open_timeout
* read_timeout
* retry_503
* retry_after_limit

### Dynamic configuration


Blacklight pre-parses the YAML file using ERB, which means you can use environment variables to configure the Solr location. With the default configuration, this means you can start Blacklight and point it at an external solr index without changing the `solr.yml` file:

```console
$ SOLR_URL="http://my.production.server/solr/core/" rails server
```

### Run-time configuration

The `solr.yml` configuration is applied globally to all Solr connections from your application. If you need to configure dynamic (e.g. per-user or per-controller) connections, Blacklight provides and uses instance-level accessor methods. Applications can override these accessors in By overriding these accessors, applications can provide either custom RSolr clients (e.g. [rsolr-async](https://github.com/mwmitchell/rsolr-async)) or per-user or per-controller solr connections:

```ruby
class CatalogController < ApplicationController
  include Blacklight::Catalog
  include MyApplicationRuntimeConfiguration
end

module MyApplicationRuntimeConfiguration
  def solr_repository
    @solr_repository ||= MyCustomSolrRepository.new(blacklight_config)
  end
end

class MyCustomSolrRepository < Blacklight::SolrRepository
  def blacklight_solr
    @blacklight_solr ||= RSolr::Custom::Client.new :user => current_user.id
  end
end
```


## Solr Configuration

### Schema


#### Solr Unique Key

If your solr configuration uses a unique field other than `id`, you must configure your `SolrDocument` (in app/models/solr_document.rb) to set the unique key:

```ruby
# app/models/solr_document.rb

class SolrDocument
  include Blacklight::Solr::Document

  self.unique_key = 'my_unique_key_field'
 
  ...
end
```

### Blacklight search requests

[Search Handler](http://wiki.apache.org/solr/SearchHandler)

Blacklight configuration parameters to directly add solr request parameters:
* default_solr_params
* qt
* solr_path
* document_solr_request_handler
* document_solr_path
* document_unique_id_param
* default_document_solr_params

### Request handlers

Blacklight supports rapid application development by allowing you to configure Blacklight to send every parameter in every solr request. One of the ways to productionize this is to move the static logic into Solr request handlers. [Request Handlers](http://wiki.apache.org/solr/SolrRequestHandler) are configured in the [solrconfig.xml](http://wiki.apache.org/solr/SolrConfigXml). 

Request handler parameters can be configured three different ways:

* defaults - provides default param values that will be used if the param does not have a value specified at request time.
* appends - provides param values that will be used in addition to any values specified at request time (or as defaults.
* invariants - provides param values that will be used in spite of any values provided at request time. They are a way of letting the Solr maintainer lock down the options available to Solr clients. Any params values specified here are used regardless of what values may be specified in either the query, the "defaults", or the "appends" params.

Here is an example request handler demonstrating all types of configuration:

```xml
  <requestHandler name="standard" class="solr.StandardRequestHandler">
     <lst name="defaults">
       <!-- assume they want 50 rows unless they tell us otherwise -->
       <int name="rows">50</int>
       <!-- assume they only want popular products unless they provide a different fq -->
       <str name="fq">popularity:[1 TO *]</str>
     </lst>
    <lst name="appends">
      <!-- no matter what other fq are also used, always restrict to only inStock products -->
      <str name="fq">inStock:true</str>
    </lst>
    <lst name="invariants">
      <!-- don't let them turn on faceting -->
      <bool name="facet">false</bool>
    </lst>

  </requestHandler>
```


#### Document request handler

In addition to the search request handler, we strongly encourage you to configure a request handler for retrieving single documents. This request handler can be highly optimized to remove unnecessary parameters and processing, makes it easier to understand the Solr request log, and allows you to easily change request parameters for search and single-item behaviors separately.

The blacklight document request handler looks like this:

```xml
<requestHandler name="document" class="solr.SearchHandler" >
  <lst name="defaults">
    <str name="echoParams">all</str>
    <str name="fl">*</str>
    <str name="rows">1</str>
    <bool name="facet">false</bool>
    <str name="q">{!raw f=id v=$id}</str> <!-- use id=666 instead of q=id:666 -->
  </lst>
</requestHandler>
```

If you're using Solr 4.0+, you may also consider using the [Solr Real-Time Get](https://cwiki.apache.org/confluence/display/solr/RealTime+Get) request handler.

```xml
<requestHandler name="/get" class="solr.RealTimeGetHandler">
  <lst name="defaults">
    <str name="omitHeader">true</str>
    <str name="wt">json</str>
    <str name="indent">true</str>
  </lst>
</requestHandler>
```

CHANGES
0c51e24 synched master README
26615aa removed strange diff output from these files
44520b2 added instructions for installing template from trunk
f1fc59a fixed path to template
c1810a9 bumped up version in template
1b96542 added VERSION and CHANGES for this branch
c62e233 updated template and readme for tag version
d64aec3 updated template and readme for tag version
d937911 updated template and readme for tag version
9fca1d8 notes about changes for v2.4.1
7bec09e updated CHANGES, which include everything from v2.4.0 to HEAD
723beb2 updated CHANGES, which include everything from v2.4.0 to HEAD
076584f changed comment
56553ac merged remote changes
1ebaca8 bumped up version in template
2cad1ab added VERSION and CHANGES for this branch
cbc7bfd updated template and readme for tag version
40fa380 updated template and readme for tag version
d10680c updated CHANGES, which include everything from v2.4.0 to HEAD
561a0c5 updated template and readme for tag version
907f40e changed comment
570c8b1 removing CHANGES and VERSION -- the 2.4-stable branch will take care of these
2d8d20c removed branch variable -- not used
dd2a40c removed old rubyforge stuff from Rakefile
15da954 woops, now adding specs for Blacklight.root and Blacklight.locate_path
4b4b6d1 added specs for Blacklight.locate_path and Blacklight.root
b39af0f added new method, Blacklight.root which returns the absolute path to the plugin install dir added new method, Blacklight.locate_path which returns a valid path, searching first in Rails.root then Blacklight.root updated solr:marc:index_test_data to use Blacklight.locate_path so plugin and apps will work with the same code
d8960f2 updated rake task and spec to work with the data submodule
1ed9a8c renamed test-data module to data
66cf4e3 changing test-data to "data" to match the template installer path
7ff7a44 set jetty dir as submodule (http://github.com/projectblacklight/blacklight-jetty)
eb5cfcf removed static jetty dir
35f6bbf updated solr:marc:index_test_data (logging to stdout) added spec for testing output of solr:marc:index_test_data task updated docs
793b02d added blacklight-data (http://github.com/projectblacklight/blacklight-data) as submodule "test-data"
d5c7f06 added solr/data/* to .gitignore
bf201da removed solr index from git
68d4115 moved cuke deps to main env file
d031944 Avoid renaming vendor/plugins/blacklight when git_export was used to pull a repository other than the plugin.  Fixes a bug where if you answer 'yes' to the 'download and configure Apache Solr' prompt from the current template.rb, it called git_export again during that step and as a side-effect the plugin was lost.
749695d updated gem deps
ce64fa1 migrated all needed files/resources to use cucumber 0.6.2 and cucumber_rails 0.2.4
35b358e Merge branch 'master' of github.com:projectblacklight/blacklight
2487dbb added VERSION file
3a8c953 Doing a little more housekeeping. Replacing more document.id with document[:id] to get rid of the Object#id will be deprecated; use Object#object_id warnings Tests seem to pass regardless of which we use.
1caef45 added rails as dependency in init.rb updated template installer:   will no longer overrwrite files   will no longer modify files if the files were modified by a previous install   saves the old installation of blacklight at vendor/{timestamp}-previous-blacklight
8cb5645 updated rspec version to 1.3.0 updated rspec-rails version to 1.3.2
d1c1f2c Merge branch 'master' of github.com:projectblacklight/blacklight
a3a0a52 printing the solr/jetty start error
fa5a0c8 Adding in Librarian View. The link is to the right of the SMS This link in the show_tools partial The link will pop up a lightbox that has the new marc_view partial rendered inside of it A new feature for the record view (CatalogController#show) has been added. The new feature checks the elements of the librarian view as well as checks some other display fields of the record.
98582f0 changing rcov options to exclude .rb files in the features directory from being counted in the coverage data
6af4dc4 removed the require "rsolr-ext" line in init.rb
b9d70c9 updated GIT docs... includes section on how to accept and merge forks
04a7ead Merge branch 'expanded_search_field_config' of git://github.com/jrochkind/blacklight into jrochkind/expanded_search_field_config
b4ab556 Accept legacy two-element Array configs in config[:search_fields]. With rspec test.
390edb5 starting place to help people with overriding parts of blacklight
5975adf updated rails dependency version to 2.3.5 removed gem dependency for mislav-will_paginate, and replaced with will_paginate version 2.3.11 - also changed its source to gemcutter.org updated rsolr-ext version to 0.11.2 - this version preserves the :fq param when using :filters and/or :phrase_filters
5eef834 Merge branch 'master' into expanded_search_field_config
95d598f spec to ensure empty string solr_parameters from search field config are used
cf9f17f document :solr_parameters key
8b7dc87 accurate comment of config[:search_fields] referencing more complete documentation
13eac1a solr_search_params looks up values from search field definition, when a :search_field is in request params.
8051529 Merge branch 'master' of github.com:jrochkind/blacklight
9b12f69 Merge branch 'master', remote branch 'origin' into expanded_search_field_config
819e3a9 in history display, don't spell out search field for default one. Default one can be specified in config, else is just first one.
da40857 search field def hash defaults to default qt if not specified
15820ff added module comments
18b64db construct solr search params using param[:search_field] to look up qt and other config in new style search field config hashes, using SearchField helpers. Still defaults to raw :qt if not found, for backwards compatibility with existing bookmarked results urls.
4a188f2 cucumber features know that param will be :search_field, a key from new style search field config hashes, not :qt

lib/blacklight/configuration.rb
module Blacklight
  ##
  # Blacklight::Configuration holds the configuration for a Blacklight::Controller, including
  # fields to display, facets to show, sort options, and search fields.
  class Configuration < OpenStructWithHashAccess

    require 'blacklight/configuration/view_config'
    require 'blacklight/configuration/tool_config'
    # XXX this isn't very pretty, but it works.
    require 'blacklight/configuration/fields'
    require 'blacklight/configuration/field'
    require 'blacklight/configuration/solr_field'
    require 'blacklight/configuration/search_field'
    require 'blacklight/configuration/facet_field'
    require 'blacklight/configuration/sort_field'
    include Fields
    extend Deprecation
    self.deprecation_horizon = 'blacklight 6.0'

    # Set up Blacklight::Configuration.default_values to contain
    # the basic, required Blacklight fields
    class << self
      def default_values

        @default_values ||= begin
          {
          ##
          # === Search request configuration
          ##
          # HTTP method to use when making requests to solr; valid
          # values are :get and :post.
          http_method: :get,
          # The solr request handler ('qt') to use for search requests
          qt: 'search',
          # The path to send requests to solr.
          solr_path: 'select',
          # Default values of parameters to send with every search request
          default_solr_params: {},
          ## deprecated; use add_facet_field :include_in_request instead;
          # if this is configured true, all facets will be included in the solr request
          # unless explicitly disabled.
          add_facet_fields_to_solr_request: false, 
          ## deprecated; use add_index_field :include_in_request instead;
          # if this is configured true, all show and index will be included in the solr request
          # unless explicitly disabled.
          add_field_configuration_to_solr_request: false,
          ##
          # === Single document request configuration
          ##
          # The solr rqeuest handler to use when requesting only a single document 
          document_solr_request_handler: 'document',
          # THe path to send single document requests to solr
          document_solr_path: nil,
          document_unique_id_param: :id,
          # Default values of parameters to send when requesting a single document
          default_document_solr_params: {
            ## Blacklight provides these settings in the /document request handler
            ## by default, we just ask for all fields. 
            #fl: '*',
            ## this is a fancy way to say "find the document by id using
            ## the value in the id query parameter"
            #q: "{!raw f=#{unique_key} v=$id}",
            ## disable features we don't need
            #facet: false,
            #rows: 1
          },
          ##
          # == Response models
          ## Class for sending and receiving requests from a search index
          repository_class: nil,
          ## Class for converting Blacklight parameters to request parameters for the repository_class
          search_builder_class: nil,
          # model that maps index responses to the blacklight response model
          response_model: nil,
          # the model to use for each response document
          document_model: nil,
          # document presenter class used by helpers and views
          document_presenter_class: nil,
          ##
          # == Blacklight view configuration
          ##
          navbar: OpenStructWithHashAccess.new(partials: { }),
          # General configuration for all views
          index: ViewConfig::Index.new(
            # solr field to use to render a document title
            title_field: nil,
            # solr field to use to render format-specific partials
            display_type_field: 'format',
            # partials to render for each document(see #render_document_partials)
            partials: [:index_header, :thumbnail, :index],
            document_actions: NestedOpenStructWithHashAccess.new(ToolConfig),
            collection_actions: NestedOpenStructWithHashAccess.new(ToolConfig),
            # what field, if any, to use to render grouped results
            group: false,
            # additional response formats for search results
            respond_to: OpenStructWithHashAccess.new()
            ),
          # Additional configuration when displaying a single document
          show: ViewConfig::Show.new(
            # default route parameters for 'show' requests
            # set this to a hash with additional arguments to merge into 
            # the route, or set `controller: :current` to route to the 
            # current controller.
            route: nil,
            # partials to render for each document(see #render_document_partials) 
            partials: [:show_header, :show],
            document_actions: NestedOpenStructWithHashAccess.new(ToolConfig)
          ),
          # Configurations for specific types of index views
          view: NestedOpenStructWithHashAccess.new(ViewConfig,
            'list',
            atom: {
              if: false, # by default, atom should not show up as an alternative view
              partials: [:document]
            },
            rss: {
              if: false, # by default, rss should not show up as an alternative view
              partials: [:document]
          }),
          #
          # These fields are created and managed below by `defined_field_access`
          # facet_fields
          # index_fields
          # show_fields
          # sort_fields
          # search_fields
          ##
          # === Blacklight behavior configuration
          ##
          # Maxiumum number of spelling suggestions to offer
          spell_max: 5,
          # Maximum number of results to show per page
          max_per_page: 100,
          # Options for the user for number of results to show per page
          per_page: [10,20,50,100],
          default_per_page: nil,
          # how many searches to save in session history
          # (TODO: move the value into the configuration?)
          search_history_window: Blacklight::Catalog::SearchHistoryWindow,
          default_facet_limit: 10
          }
        end
      end
    end

    ##
    # Create collections of solr field configurations.
    # This will create array-like accessor methods for
    # the given field, and an #add_x_field convenience 
    # method for adding new fields to the configuration
    
    # facet fields
    define_field_access :facet_field
    
    # solr fields to display on search results
    define_field_access :index_field
    
    # solr fields to display when showing single documents
    define_field_access :show_field
    
    # solr "fields" to use for scoping user search queries
    # to particular fields
    define_field_access :search_field
    
    # solr fields to use for sorting results
    define_field_access :sort_field

    def initialize(*args)
      super(*args)
      initialize_default_values!
      yield(self) if block_given?
      self
    end

    ##
    # Initialize default values from the class attribute
    def initialize_default_values!
      Marshal.load(Marshal.dump(self.class.default_values)).each do |k, v|
        self[k] ||=  v
      end
    end
    
    def document_model
      super || SolrDocument
    end
    alias_method :solr_document_model, :document_model
    # only here to support alias_method
    def document_model= *args
      super
    end
    alias_method :solr_document_model=, :document_model=
    deprecation_deprecate solr_document_model: :document_model, :solr_document_model= => :document_model=

    def document_presenter_class
      super || Blacklight::DocumentPresenter
    end

    def response_model
      super || Blacklight::SolrResponse
    end
    alias_method :solr_response_model, :response_model
    # only here to support alias_method
    def response_model= *args
      super
    end
    alias_method :solr_response_model=, :response_model=
    deprecation_deprecate solr_response_model: :response_model, :solr_response_model= => :response_model=

    def repository_class
      super || Blacklight::SolrRepository
    end

    def search_builder_class
      super || Blacklight::Solr::SearchBuilder
    end

    def default_per_page
      super || per_page.first
    end

    ##
    # DSL helper
    def configure
      yield self if block_given?
      self
    end

    ##
    # Returns default search field, used for simpler display in history, etc.
    # if not set, defaults to first defined search field
    def default_search_field
      field = super
      field ||= search_fields.values.select { |field| field.default == true }.first
      field ||= search_fields.values.first

      field
    end

    ##
    # Returns default sort field, used for simpler display in history, etc.
    # if not set, defaults to first defined sort field
    def default_sort_field
      field = super
      field ||= sort_fields.values.select { |field| field.default == true }.first
      field ||= sort_fields.values.first

      field
    end
    
    def default_title_field
      document_model.unique_key || 'id'
    end

    ##
    # Add any configured facet fields to the default solr parameters hash
    # @overload add_facet_fields_to_solr_request!
    #    add all facet fields to the solr request
    # @overload add_facet_fields_to_solr_request! field, field, field
    #   @param [Symbol] Field names to add to the solr request
    #   @param [Symbol] 
    #   @param [Symbol] 
    def add_facet_fields_to_solr_request! *fields
      if fields.empty?
        self.add_facet_fields_to_solr_request = true
      else
        facet_fields.slice(*fields).each do |k,v|
          v.include_in_request = true
        end
      end
    end

    ##
    # Add any configured facet fields to the default solr parameters hash
    # @overload add_field_configuration_to_solr_request!
    #    add all index, show, and facet fields to the solr request
    # @overload add_field_configuration_to_solr_request! field, field, field
    #   @param [Symbol] Field names to add to the solr request
    #   @param [Symbol] 
    #   @param [Symbol] 
    def add_field_configuration_to_solr_request! *fields
      if fields.empty?
        self.add_field_configuration_to_solr_request = true
      else
        index_fields.slice(*fields).each do |k,v|
          v.include_in_request = true
        end

        show_fields.slice(*fields).each do |k,v|
          v.include_in_request = true
        end
        facet_fields.slice(*fields).each do |k,v|
          v.include_in_request = true
        end
      end
    end

    ##
    # Deprecated. Get the list of facet fields to explicitly
    # add to the solr request
    def facet_fields_to_add_to_solr
      facet_fields.select { |k,v| v.include_in_request }
                  .reject { |k,v| v[:query] || v[:pivot] }
                  .map { |k,v| v.field }
    end
    deprecation_deprecate :facet_fields_to_add_to_solr

    ##
    # Provide a 'deep copy' of Blacklight::Configuration that can be modifyed without affecting
    # the original Blacklight::Configuration instance.
    #
    # The Rails 3.x version only copies hashes, and ignores arrays and similar structures
    if ::Rails.version < "4.0"
      def deep_copy
        Marshal.load(Marshal.dump(self))
      end
      alias_method :inheritable_copy, :deep_copy
    else
      ##
      # Rails 4.x provides `#deep_dup`, but it aggressively `#dup`'s class names
      # too. These model names should not be `#dup`'ed or we might break ActiveModel::Naming.
      def deep_copy
        deep_dup.tap do |copy|
          copy.repository_class = self.repository_class
          copy.response_model = self.response_model
          copy.document_model = self.document_model
          copy.document_presenter_class = self.document_presenter_class
          copy.search_builder_class = self.search_builder_class
        end
      end
      alias_method :inheritable_copy, :deep_copy
    end

    ##
    # Get a view configuration for the given view type
    # including default values from the index configuration
    def view_config view_type
      if view_type == :show
        self.index.merge self.show
      else
        self.index.merge view.fetch(view_type, {})
      end
    end

    ##
    # Add a partial to the tools when rendering a document.
    # @param partial [String] the name of the document partial
    # @param opts [Hash]
    # @option opts [Symbol,Proc] :if render this action if the method identified by the symbol or the proc evaluates to true.
    #                             The proc will receive the action configuration and the document or documents for the action.
    # @option opts [Symbol,Proc] :unless render this action unless the method identified by the symbol or the proc evaluates to true
    #                             The proc will receive the action configuration and the document or documents for the action.
    def add_show_tools_partial(name, opts = {})
      opts[:partial] ||= 'document_action'
      add_action(show.document_actions, name, opts)
    end

    ##
    # Add a tool for the search result list itself
    # @param partial [String] the name of the document partial
    # @param opts [Hash]
    # @option opts [Symbol,Proc] :if render this action if the method identified by the symbol or the proc evaluates to true.
    #                             The proc will receive the action configuration and the document or documents for the action.
    # @option opts [Symbol,Proc] :unless render this action unless the method identified by the symbol or the proc evaluates to true
    #                             The proc will receive the action configuration and the document or documents for the action.
    def add_results_collection_tool(name, opts = {})
      add_action(index.collection_actions, name, opts)
    end

    ##
    # Add a partial to the tools for each document in the search results.
    # @param partial [String] the name of the document partial
    # @param opts [Hash]
    # @option opts [Symbol,Proc] :if render this action if the method identified by the symbol or the proc evaluates to true.
    #                             The proc will receive the action configuration and the document or documents for the action.
    # @option opts [Symbol,Proc] :unless render this action unless the method identified by the symbol or the proc evaluates to true
    #                             The proc will receive the action configuration and the document or documents for the action.
    def add_results_document_tool(name, opts = {})
      add_action(index.document_actions, name, opts)
    end

    ##
    # Add a partial to the header navbar
    # @param partial [String] the name of the document partial
    # @param opts [Hash]
    # @option opts [Symbol,Proc] :if render this action if the method identified by the symbol or the proc evaluates to true.
    #                             The proc will receive the action configuration and the document or documents for the action.
    # @option opts [Symbol,Proc] :unless render this action unless the method identified by the symbol or the proc evaluates to true
    #                             The proc will receive the action configuration and the document or documents for the action.
    def add_nav_action name, opts = {}
      add_action(navbar.partials, name, opts)
    end

    private

      def add_action config_hash, name, opts
        config = Blacklight::Configuration::ToolConfig.new opts
        config.name = name

        if block_given?
          yield config
        end

        config_hash[name] = config
      end
  end
end
